core/drivers/stm32_hpdma.c

```diff
@@ -0,0 +1,349 @@
+
+/*
+ * SECCFGR register bitfields
+ */
+#define _HPDMA_SECCFGR_EN		BIT(0)
+#define _HPDMA_SECCFGR_MASK		GENMASK_32(15, 0)
+
+/*
+ * SEMCR register bitfields
+ */
+#define _HPDMA_SEMCR_SCID_MASK		GENMASK_32(5, 4)
+#define _HPDMA_SEMCR_SCID_SHIFT		U(4)
+
+/*
+ * Miscellaneous
+ */
+
+#define HPDMA_RIF_CHANNELS		U(16)
+
+#define HPDMA_NB_MAX_CID_SUPPORTED	U(3)
+
+struct hpdma_pdata {
+	struct clk *hpdma_clock;
+	struct rif_conf_data conf_data;
+	unsigned int nb_channels;
+	vaddr_t base;
+
+	SLIST_ENTRY(hpdma_pdata) link;
+};
+
+static SLIST_HEAD(, hpdma_pdata) hpdma_list =
+		SLIST_HEAD_INITIALIZER(hpdma_list);
+
+/* This function expects HPDMA bus clock is enabled */
+static TEE_Result apply_rif_config(struct hpdma_pdata *hpdma_d, bool is_tdcid)
+{
+	TEE_Result res = TEE_ERROR_ACCESS_DENIED;
+	uint32_t cidcfgr = 0;
+	unsigned int i = 0;
+
+	for (i = 0; i < HPDMA_RIF_CHANNELS; i++) {
+		if (!(BIT(i) & hpdma_d->conf_data.access_mask[0]))
+			continue;
+		/*
+		 * When TDCID, OP-TEE should be the one to set the CID filtering
+		 * configuration. Clearing previous configuration prevents
+		 * undesired events during the only legitimate configuration.
+		 */
+		if (is_tdcid)
+			io_clrbits32(hpdma_d->base + _HPDMA_CIDCFGR(i),
+				     _HPDMA_CIDCFGR_CONF_MASK);
+
+		cidcfgr = io_read32(hpdma_d->base + _HPDMA_CIDCFGR(i));
+
+		/* Check if the channel is in semaphore mode */
+		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1))
+			continue;
+
+		/* If not TDCID, we want to acquire semaphores assigned to us */
+		res = stm32_rif_acquire_semaphore(hpdma_d->base +
+						  _HPDMA_SEMCR(i),
+						  HPDMA_NB_MAX_CID_SUPPORTED);
+		if (res) {
+			EMSG("Couldn't acquire semaphore for channel %u", i);
+			return res;
+		}
+	}
+	/* Security and privilege RIF configuration */
+	io_clrsetbits32(hpdma_d->base + _HPDMA_PRIVCFGR, _HPDMA_PRIVCFGR_MASK &
+			hpdma_d->conf_data.access_mask[0],
+			hpdma_d->conf_data.priv_conf[0]);
+	io_clrsetbits32(hpdma_d->base + _HPDMA_SECCFGR, _HPDMA_SECCFGR_MASK &
+			hpdma_d->conf_data.access_mask[0],
+			hpdma_d->conf_data.sec_conf[0]);
+
+	/* Skip CID/semaphore configuration if not in TDCID state. */
+	if (!is_tdcid)
+		goto end;
+
+	for (i = 0; i < HPDMA_RIF_CHANNELS; i++) {
+		if (!(BIT(i) & hpdma_d->conf_data.access_mask[0]))
+			continue;
+
+		io_clrsetbits32(hpdma_d->base + _HPDMA_CIDCFGR(i),
+				_HPDMA_CIDCFGR_CONF_MASK,
+				hpdma_d->conf_data.cid_confs[i]);
+
+		cidcfgr = io_read32(hpdma_d->base + _HPDMA_CIDCFGR(i));
+
+		/*
+		 * Take semaphore if the resource is in semaphore
+		 * mode and secured.
+		 */
+		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1) ||
+		    !(io_read32(hpdma_d->base + _HPDMA_SECCFGR) & BIT(i))) {
+			res =
+			stm32_rif_release_semaphore(hpdma_d->base +
+						    _HPDMA_SEMCR(i),
+						    HPDMA_NB_MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Couldn't release semaphore for res%u", i);
+				return TEE_ERROR_ACCESS_DENIED;
+			}
```

Comment:
here, if `res == TEE_SUCCESS`, we still failed to aquire the semaphore (test at line 150). Shouldn't this also return a error code?

Reply:
Mmm i don't think so, the goal here is to release any semaphore that could have been taken by the CID1 if it's not in the authorized list anymore.

Reply:
ok